update resource metrics pipeline section (#31558)
kubernetes/website: content/en/docs/tasks/debug-application-cluster/resource-metrics-pipeline.md

---
reviewers:

> - fgrzadkowski

Reviewer: If we are updating documentation, we should also update reviewers. Both reviewers currently listed haven't contributed in last 3 years.

Reviewer 2: Sounds good to me.

Reviewer 2: @chrismetz09 in my view it's also reasonable to update the page reviewers in a separate, follow-up PR

> - piosz 

Reviewer: Please avoid add a tailing whitespaces.

title: Resource metrics pipeline 
content_type: concept
---

<!-- overview -->

> Kubernetes supports a built-in resource metrics pipeline. A fixed set of metrics consisting of node and pod CPU and memory usage are gathered up by the metrics-server and exposed to other components across the Metrics API.

Reviewer: ```suggestion
Kubernetes supports a built-in resource metrics pipeline. A fixed set of metrics consisting of node and pod CPU and memory usage for Kubernetes built-in workload autoscaling pipelines. 
```

{{< note >}}
The resource metrics pipeline only supports CPU and memory metrics for use by HPA and VPA. To access a more complete set of metrics, you will need to deploy a [Full metrics pipeline](https://kubernetes.io/docs/tasks/debug-application-cluster/resource-usage-monitoring/#full-metrics-pipeline). 
{{< /note >}}

Figure 1 illustrates the architecture of the resource metrics pipeline.

{{< mermaid >}}
flowchart RL
subgraph cluster[Cluster]
direction RL
S[ <br><br> ]
A[Metrics-<br>Server]
subgraph B[Nodes]
direction TB
D[cAdvisor] --> C[kubelet]
E[Container<br>runtime] --> D
E1[Container<br>runtime] --> D
P[pod data] -.- C
end
L[API<br>server]
W[HPA]
C ---->|Summary<br>API| A -->|metrics<br>API| L --> W
end
L ---> K[kubectl<br>top]
classDef box fill:#fff,stroke:#000,stroke-width:1px,color:#000;
class W,B,P,K,cluster,D,E,E1 box
classDef spacewhite fill:#ffffff,stroke:#fff,stroke-width:0px,color:#000
class S spacewhite
classDef k8s fill:#326ce5,stroke:#fff,stroke-width:1px,color:#fff;
class A,L,C k8s
{{< /mermaid >}}

Figure 1. Resource Metrics Pipeline

The architecture components, from right to left in the figure, consist of the following:

> * [cAdvisor](https://github.com/google/cadvisor) - Daemon for collecting, aggregating and exposing Docker container metrics. Note that non-Docker container runtimes that support metrics collection through a [CRI will not require cAdvisor](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-node/cri-container-stats.md).

Reviewer: ```suggestion
* [cAdvisor](https://github.com/google/cadvisor) - Daemon for collecting, aggregating and exposing container metrics compiled into Kubelet. cAdvisor reads metrics from cgroups allowing out of the box support for Docker. Note that non-container runtimes need to support [Container Metrics RPS](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-node/cri-container-stats.md) for metrics to be available.
```

> * [kubelet](https://kubernetes.io/docs/concepts/overview/components/#kubelet) - Node agent for managing container resources. Resource metrics are accessible using the `/nodes/resource` and `/nodes/stats` kubelet API endpoints. 

Reviewer 3: ```suggestion
* [kubelet](https://kubernetes.io/docs/concepts/overview/components/#kubelet) - Node agent for managing container resources. Resource metrics are accessible using the `/metrics/resource` and `/stats` kubelet API endpoints. 
```

> * [kubelet](https://kubernetes.io/docs/concepts/overview/components/#kubelet) - Node agent for managing container resources. Resource metrics are accessible using the `/nodes/resource` and `/nodes/stats` kubelet API endpoints. 

Reviewer: ```suggestion
* [kubelet](https://kubernetes.io/docs/concepts/overview/components/#kubelet) - Node agent for managing containers. Reads metrics from either cAdvisor or CRI and exposes them via `/nodes/resource` and `/nodes/stats` (deprcated) kubelet API endpoints. 
```

> * [Summary API](#summary-api-source) - Kubelet API for discovering and retrieving per-node summarized stats available through the `/nodes/stats` endpoint. 

Reviewer 3: ```suggestion
* [Summary API](#summary-api-source) - Kubelet API for discovering and retrieving per-node summarized stats available through the `/stats` endpoint. 
```

> * [metrics-server](#metrics-server) - Cluster addon component that collects and aggregates resource metrics pulled from each kubelet. The K8s API server serves up the Metric API endpoints for use by HPA, VPA and the `kubectl top` command.

Reviewer: ```suggestion
* [metrics-server](#metrics-server) - Cluster addon component that collects and aggregates resource metrics pulled from each kubelet. Metrics Server is the default implementation of Metric API used by HPA, VPA and the `kubectl top` command.
```

* [Metrics API](#metrics-api) - Kubernetes API supporting access to CPU and memory metrics provided by the metrics-server. 
  
<!-- body -->

## Metrics API

The metrics-server implements the Metrics API. This API allows you to access CPU and memory usage for the nodes and pods in your cluster. Its primary role is to feed resource usage metrics to K8s autoscaler components. 

Here is an example of the Metrics API request for a `minikube` node piped through `jq` for easier reading:
```text
kubectl get --raw "/apis/metrics.k8s.io/v1beta1/nodes/minikube" | jq '.'
```
Here is the same API call using `curl`:
```text

> curl http://localhost:8080/apis/metrics.k8s.io/v1beta1/nodes/minikube

Reviewer: This assumes that there is `kubectl proxy` running in background

Author: Used minikube. Your recommendation to:
1. add note indicating examples use minikube and kubectl proxy.
2. do nothing

Reviewer: I think we should not assume any specific K8s distro in documentation. Please update this to recommend forwarding local port to K8s apiserver by running `kubectl proxy`.

```text
kubectl get --raw "/apis/metrics.k8s.io/v1beta1/namespaces/kube-system/pods/kube-scheduler-minikube" | jq '.'
```
Here is the same API call using `curl`:
```text
curl http://localhost:8080/apis/metrics.k8s.io/v1beta1/namespaces/kube-system/pods/kube-scheduler-minikube
```
Sample reply:
```text
{
  "kind": "PodMetrics",
  "apiVersion": "metrics.k8s.io/v1beta1",
  "metadata": {
    "name": "kube-scheduler-minikube",
    "namespace": "kube-system",
    "selfLink": "/apis/metrics.k8s.io/v1beta1/namespaces/kube-system/pods/kube-scheduler-minikube",
    "creationTimestamp": "2022-01-27T19:25:00Z"
  },
  "timestamp": "2022-01-27T19:24:31Z",
  "window": "30s",
  "containers": [
    {
      "name": "kube-scheduler",
      "usage": {
        "cpu": "9559630n",
        "memory": "22244Ki"
      }
    }
  ]
}
```

The Metrics API is defined in the [k8s.io/metrics](https://github.com/kubernetes/metrics) repository. You must enable the [API aggregation layer](https://kubernetes.io/docs/tasks/extend-kubernetes/configure-aggregation-layer/) and register the `metrics.k8s.io` API. 

> To learn more about the Metrics API, see [resource metrics API design](https://github.com/kubernetes/design-proposals-archive/blob/main/instrumentation/resource-metrics-api.md), the [metrics repository](https://github.com/kubernetes-sigs/metrics-server) and the [resource metrics API](https://github.com/kubernetes/metrics#resource-metrics-api).

Reviewer 3: ```suggestion
To learn more about the Metrics API, see [resource metrics API design](https://github.com/kubernetes/design-proposals-archive/blob/main/instrumentation/resource-metrics-api.md), the [metrics-server repository](https://github.com/kubernetes-sigs/metrics-server) and the [resource metrics API](https://github.com/kubernetes/metrics#resource-metrics-api).
```

{{< note >}} You must deploy the metrics-server in your cluster to access the Metrics API. {{< /note >}}

> ## Measuring Resource Usage

Reviewer 2: extra cleanup opportunity:
```suggestion
## Measuring resource usage
```

In an ideal world, the "working set" is the amount of memory in-use that cannot be freed under memory pressure. However, calculation of the working set varies by host OS, and generally makes heavy use of heuristics to produce an estimate. 

It includes all anonymous (non-file-backed) memory since Kubernetes does not support memory swap. The metric also includes some cached (file-backed) memory, because the host OS cannot always reclaim pages.

To learn more about memory, see the [Meaning of memory](https://kubernetes.io/docs/concepts/configuration/manage-compute-resources-container/#meaning-of-memory).

## Metrics Server

The metrics-server fetches resource metrics from the kubelets and exposes them in the Kubernetes API server through the Metrics API for use by the HPA and VPA. You can also view these metrics using the `kubectl top` command.

The metrics-server watches for all nodes and pods in your cluster. Nodes for metrics collection discovery; pods for metadata. It uses pod metadata to check pod readiness, and to handle queries with label selectors. 

For example with an HPA query, the metrics-server needs to identify which pods fulfill the label selectors in the deployment.

> The metrics-server employs the Summary API to fetch metrics exposed by
> [kubelet](/docs/reference/command-line-tools-reference/kubelet/) on each node.

Reviewer: ```suggestion
The metrics-server calls 
[kubelet](/docs/reference/command-line-tools-reference/kubelet/) API to collect metrics from each node. Depending on metrics-server version it uses:
* Metrics resource endpoint `/metrics/resource` in version v0.6.0+ or
* Summary API endpoint `/stats/summary` in older versions

```

To learn more about the metrics-server, see the [metrics-server repository](https://github.com/kubernetes-sigs/metrics-server).

You can also check out the following:

* [metrics-server design](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/instrumentation/metrics-server.md)
* [metrics-server FAQ](https://github.com/kubernetes-sigs/metrics-server/blob/master/FAQ.md)
* [metrics-server known issues](https://github.com/kubernetes-sigs/metrics-server/blob/master/KNOWN_ISSUES.md)
* [metrics-server releases](https://github.com/kubernetes-sigs/metrics-server/releases)

> * [HorizontalPodAutoscaler (HPA)](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/)

Reviewer 2: ```suggestion
* [Horizontal Pod Autoscaling](/docs/tasks/run-application/horizontal-pod-autoscale/)
```

> ### Summary API Source

Reviewer 2: ```suggestion
### Summary API source
```

> ### Summary API source

Reviewer: As this endpoint is deprecated we should consider suggesting documenting Resource Metrics endpoint instead `/metrics/resource`

Reviewer 2: We usually document deprecated features. We can (and should) mark things deprecated to make that clear, ideally with an indication of the release of Kubernetes that deprecated it.

The [kubelet](/docs/reference/command-line-tools-reference/kubelet/) gathers stats at the node, volume, pod and container level, and emits this information in
the [Summary API](https://github.com/kubernetes/kubernetes/blob/7d309e0104fedb57280b261e5677d919cb2a0e2d/staging/src/k8s.io/kubelet/pkg/apis/stats/v1alpha1/types.go)
for consumers to read.

Here is an example of a Summary API request for a `minikube` node:
```text
kubectl get --raw /api/v1/nodes/minikube/proxy/stats/summary
```
Here is the same API call using `curl`:
```text
curl http://localhost:8080/api/v1/nodes/minikube/proxy/stats/summary
```

{{< note >}}

> Metrics-server v0.6.0 replaces the summary API `/nodes/stats/` endpoint with the kubelet API `/nodes/metrics/resource` endpoint for pulling stats. For more details, see [metrics-server releases](https://github.com/kubernetes-sigs/metrics-server/releases).

Reviewer 3: ```suggestion
Metrics-server v0.6.0 replaces the summary API `/stats` endpoint with the kubelet API `/metrics/resource` endpoint for pulling stats. For more details, see [metrics-server releases](https://github.com/kubernetes-sigs/metrics-server/releases).
```

{{< /note >}}

> Pre-K8s 1.23, these resources have been primarily gathered from [cAdvisor](https://github.com/google/cadvisor). However, in 1.23 with the introduction of the `PodAndContainerStatsFromCRI` FeatureGate, container and pod level stats can be gathered by the CRI implementation. 

Reviewer: Not sure if this is correct. I think the pre 1.23 behavior was based on fallback. cAdvisor -> CRI. I expect that this feature gate changed the preference to pick CRI over cAdvisor.

The [HorizontalPodAutoscaler (HPA)](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/) and [VerticalPodAutoscaler (VPA)](https://github.com/kubernetes/autoscaler/tree/master/vertical-pod-autoscaler#readme) use these metrics to adjust cluster resources to meet your workload demand.

You can also view the resource metrics using the [`kubectl top`](https://kubernetes.io/docs/reference/generated/kubectl/kubectl-commands#top) command.

{{< note >}}
The resource metrics pipeline only supports CPU and memory metrics for use by HPA and VPA. To access a more complete set of metrics, you will need to deploy a [Full metrics pipeline](https://kubernetes.io/docs/tasks/debug-application-cluster/resource-usage-monitoring/#full-metrics-pipeline). 
{{< /note >}}

Figure 1 illustrates the architecture of the resource metrics pipeline.

{{< mermaid >}}
flowchart TB
subgraph cluster[Cluster]
direction RL
S[ <br><br> ]
A[metrics-<br>server]
subgraph B[Nodes]
direction TB
D[cAdvisor] --> C[kubelet]
E[Container<br>runtime] --> D
E1[Container<br>runtime] --> D
P[pod data] -.- C
end
L[API<br>server]
W[HPA]

> K[kubectl<br>Top]

Reviewer 2: (important) nit: `kubectl` should be shown as outside the cluster

---
reviewers:
- fgrzadkowski
- piosz 
title: Resource metrics pipeline 
content_type: concept
---

<!-- overview -->

> Kubernetes supports a built-in resource metrics pipeline. A fixed set of metrics consisting of node and pod CPU and memory usage for Kubernetes built-in workload autoscaling pipelines. 

Reviewer 2: ```suggestion
For Kubernetes, the _Metrics API_ offers a basic set of metrics to support automatic scaling and similar use cases.
This API makes information available about resource usage for node and pod, including metrics for CPU and memory.
If you deploy the Metrics API into your cluster, clients of the Kubernetes API can then query for this information, and
you can use Kubernetes' access control mechanisms to manage permissions to do so.
```
?

> The [HorizontalPodAutoscaler (HPA)](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/) and [VerticalPodAutoscaler (VPA)](https://github.com/kubernetes/autoscaler/tree/master/vertical-pod-autoscaler#readme) use these metrics to adjust workload replicas and resources to meet customer demand.

Reviewer 2: ```suggestion
The [HorizontalPodAutoscaler](/docs/tasks/run-application/horizontal-pod-autoscale/)  (HPA) and [VerticalPodAutoscaler](https://github.com/kubernetes/autoscaler/tree/master/vertical-pod-autoscaler#readme) (VPA) use data from the metrics API to adjust workload replicas and resources to meet customer demand.
```

You can also view the resource metrics using the [`kubectl top`](https://kubernetes.io/docs/reference/generated/kubectl/kubectl-commands#top) command.

{{< note >}}

> The resource metrics pipeline only supports CPU and memory metrics for use by HPA and VPA. To access a more complete set of metrics, you will need to deploy a [Full metrics pipeline](https://kubernetes.io/docs/tasks/debug-application-cluster/resource-usage-monitoring/#full-metrics-pipeline). 

Reviewer 2: ```suggestion
The Metrics API, and the metrics pipeline that it enables, only offers the minimum
CPU and memory metrics to enable automatic scaling using HPA and / or VPA.
If you would like to provide a more complete set of metrics, you can complement
the simpler Metrics API by deploying a second
[metrics pipeline](/docs/tasks/debug-application-cluster/resource-usage-monitoring/#full-metrics-pipeline)
that uses the _Custom Metrics API_.
```

Reviewer: `If you would like to provide a more complete set of metrics` -> `If you would like to autoscale based on other metrics`

{{< /note >}}

Figure 1 illustrates the architecture of the resource metrics pipeline.

{{< mermaid >}}
flowchart RL
subgraph cluster[Cluster]
direction RL
S[ <br><br> ]
A[Metrics-<br>Server]
subgraph B[Nodes]
direction TB
D[cAdvisor] --> C[kubelet]
E[Container<br>runtime] --> D
E1[Container<br>runtime] --> D
P[pod data] -.- C
end
L[API<br>server]
W[HPA]
C ---->|Summary<br>API| A -->|metrics<br>API| L --> W
end
L ---> K[kubectl<br>top]
classDef box fill:#fff,stroke:#000,stroke-width:1px,color:#000;
class W,B,P,K,cluster,D,E,E1 box
classDef spacewhite fill:#ffffff,stroke:#fff,stroke-width:0px,color:#000
class S spacewhite
classDef k8s fill:#326ce5,stroke:#fff,stroke-width:1px,color:#fff;
class A,L,C k8s
{{< /mermaid >}}

Figure 1. Resource Metrics Pipeline

The architecture components, from right to left in the figure, consist of the following:

> * [cAdvisor](https://github.com/google/cadvisor) - Daemon for collecting, aggregating and exposing Docker container metrics. Note that non-Docker container runtimes that support metrics collection through a [CRI will not require cAdvisor](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-node/cri-container-stats.md).
> * [kubelet](https://kubernetes.io/docs/concepts/overview/components/#kubelet) - Node agent for managing container resources. Resource metrics are accessible using the `/metrics/resource` and `/stats` kubelet API endpoints. 
> * [Summary API](#summary-api-source) - Kubelet API for discovering and retrieving per-node summarized stats available through the `/stats` endpoint. 
> * [metrics-server](#metrics-server) - Cluster addon component that collects and aggregates resource metrics pulled from each kubelet. The K8s API server serves up the Metric API endpoints for use by HPA, VPA and the `kubectl top` command.
> * [Metrics API](#metrics-api) - Kubernetes API supporting access to CPU and memory used for workload autoscaling. Metrics Server is the default implementation provided with many popular K8s distributions, however it can be replaced by alternative adapters based on your preferred monitoring solution.

Reviewer 2: _**not checked for technical accuracy**_ and probably OK to save for a follow-up PR.
```suggestion
[cAdvisor](https://github.com/google/cadvisor)
: daemon for collecting, aggregating and exposing container metrics included in Kubelet. 
* [kubelet](/docs/concepts/overview/components/#kubelet)
: Node agent for managing container resources. Resource metrics are accessible using the `/metrics/resource` and `/stats` kubelet API endpoints. 
* [Summary API](#summary-api-source)
: API provided by the kubelet for discovering and retrieving per-node summarized stats available through the `/stats` endpoint. 
* [metrics-server](#metrics-server)
: Cluster addon component that collects and aggregates resource metrics pulled from each kubelet. The API server serves Metrics API for use by HPA, VPA, and by the `kubectl top` command. Metrics Server is a reference implementation of the Metrics API.
* [Metrics API](#metrics-api)
: Kubernetes API supporting access to CPU and memory used for workload autoscaling. To make this work in your cluster, you need an API extension server that provides the Metrics API. 

{{< note >}}
cAdvisor supports reading metrics from cgroups, which works with typical container runtimes on Linux.

If you use a container runtime that uses another resource isolation mechanism, for example virtualization, then that container runtime must support [CRI Container Metrics](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-node/cri-container-stats.md) in order for metrics to be available to the kubelet.
{{< /note >}}
```
Let's be vendor-agnostic about container runtime. I also recognize that we might need to revise this again for v1.24

Author: commit suggestion greyed out. will add in separate commit

  
<!-- body -->

## Metrics API

The metrics-server implements the Metrics API. This API allows you to access CPU and memory usage for the nodes and pods in your cluster. Its primary role is to feed resource usage metrics to K8s autoscaler components. 

Here is an example of the Metrics API request for a `minikube` node piped through `jq` for easier reading:

> ```text

Reviewer 2: nit: shell, not text. You could also use two tabs (`kubectl` and `curl`).

kubectl get --raw "/apis/metrics.k8s.io/v1beta1/nodes/minikube" | jq '.'
```
Here is the same API call using `curl`:

> ```text

Reviewer 2: nit: shell, not text

curl http://localhost:8080/apis/metrics.k8s.io/v1beta1/nodes/minikube
```
Sample reply:

> ```text

Reviewer 2: nit: this is JSON, not text.

Author: will do in separate commit to test locally

{
  "kind": "NodeMetrics",

>   "apiVersion": "metrics.k8s.io/v1beta1",

Reviewer 2: BTW, if this API is still beta then we should document somewhere in this page that it's not yet a stable feature. That can happen in a follow up PR; this PR already has lots of changes.

Author: okay, future PR it is. Will track in new issue

  "metadata": {
    "name": "minikube",
    "selfLink": "/apis/metrics.k8s.io/v1beta1/nodes/minikube",
    "creationTimestamp": "2022-01-27T18:48:43Z"
  },
  "timestamp": "2022-01-27T18:48:33Z",
  "window": "30s",
  "usage": {
    "cpu": "487558164n",
    "memory": "732212Ki"
  }
}
```
Here is an example of the Metrics API request for a `kube-scheduler-minikube` pod contained in the `kube-system` namespace and piped through `jq` for easier reading:

```text
kubectl get --raw "/apis/metrics.k8s.io/v1beta1/namespaces/kube-system/pods/kube-scheduler-minikube" | jq '.'
```
Here is the same API call using `curl`:
```text
curl http://localhost:8080/apis/metrics.k8s.io/v1beta1/namespaces/kube-system/pods/kube-scheduler-minikube
```
Sample reply:

> ```text

Reviewer 2: The syntax highlighting is missing here. We should have it (by marking the blocks as shell, JSON, etc).

Author: will do in separate commit to test locally

{
  "kind": "PodMetrics",
  "apiVersion": "metrics.k8s.io/v1beta1",
  "metadata": {
    "name": "kube-scheduler-minikube",
    "namespace": "kube-system",
    "selfLink": "/apis/metrics.k8s.io/v1beta1/namespaces/kube-system/pods/kube-scheduler-minikube",
    "creationTimestamp": "2022-01-27T19:25:00Z"
  },
  "timestamp": "2022-01-27T19:24:31Z",
  "window": "30s",
  "containers": [
    {
      "name": "kube-scheduler",
      "usage": {
        "cpu": "9559630n",
        "memory": "22244Ki"
      }
    }
  ]
}
```

> The Metrics API is defined in the [k8s.io/metrics](https://github.com/kubernetes/metrics) repository. You must enable the [API aggregation layer](https://kubernetes.io/docs/tasks/extend-kubernetes/configure-aggregation-layer/) and register the `metrics.k8s.io` API. 

Reviewer 2: nit:
```suggestion
The Metrics API is defined in the [k8s.io/metrics](https://github.com/kubernetes/metrics) repository. You must enable the [API aggregation layer](/docs/tasks/extend-kubernetes/configure-aggregation-layer/) and register an [APIService](/docs/reference/kubernetes-api/cluster-resources/api-service-v1/) for the `metrics.k8s.io` API.
```

To learn more about the Metrics API, see [resource metrics API design](https://github.com/kubernetes/design-proposals-archive/blob/main/instrumentation/resource-metrics-api.md), the [metrics-server repository](https://github.com/kubernetes-sigs/metrics-server) and the [resource metrics API](https://github.com/kubernetes/metrics#resource-metrics-api).

{{< note >}} You must deploy the metrics-server or alternative adapter that serves the Metrics API to be able to access it. {{< /note >}}

## Measuring resource usage

### CPU

CPU is reported as the average core usage measured in cpu units. One cpu, in Kubernetes, is equivalent to 1 vCPU/Core for cloud providers, and 1 hyper-thread on bare-metal Intel processors.

This value is derived by taking a rate over a cumulative CPU counter provided by the kernel (in both Linux and Windows kernels). The time window used to calculate CPU is shown under window field in Metrics API.

> To learn more about the CPU, see [Meaning of CPU](https://kubernetes.io/docs/concepts/configuration/manage-compute-resources-container/#meaning-of-cpu).

Reviewer 2: ```suggestion
To learn more about how Kubernetes allocates and measures CPU resources, see [meaning of CPU](/docs/concepts/configuration/manage-compute-resources-container/#meaning-of-cpu).
```

### Memory

Memory is reported as the working set, measured in bytes, at the instant the metric was collected.

In an ideal world, the "working set" is the amount of memory in-use that cannot be freed under memory pressure. However, calculation of the working set varies by host OS, and generally makes heavy use of heuristics to produce an estimate. 

> It includes all anonymous (non-file-backed) memory since Kubernetes does not support memory swap. The metric also includes some cached (file-backed) memory, because the host OS cannot always reclaim pages.

Reviewer 2: If editing this line, it should be accurate. Kubernetes has [alpha support for swap](https://kubernetes.io/docs/concepts/architecture/nodes/#swap-memory). I don't know how that alpha feature interacts with container runtimes and the way they report memory use.

```suggestion
The Kubernetes model for a container's working set expects that the container runtime counts anonymous memory associated with the container in question. The working set metric typically also includes some cached (file-backed) memory, because the host OS cannot always reclaim pages.
```

> To learn more about memory, see the [Meaning of memory](https://kubernetes.io/docs/concepts/configuration/manage-compute-resources-container/#meaning-of-memory).

Reviewer 2: ```suggestion
To learn more about how Kubernetes allocates and measures memory resources, see [meaning of memory](/docs/concepts/configuration/manage-compute-resources-container/#meaning-of-memory).
```
(although, since the definition we're linking to unsurprising in terms of metrics, it might be OK to omit this paragraph).

## Metrics Server

The metrics-server fetches resource metrics from the kubelets and exposes them in the Kubernetes API server through the Metrics API for use by the HPA and VPA. You can also view these metrics using the `kubectl top` command.

> The metrics-server watches for all nodes and pods in your cluster. Nodes for metrics collection discovery; pods for metadata. It uses pod metadata to check pod readiness, and to handle queries with label selectors. 

Reviewer 2: Is this correct?
```suggestion
The metrics-server uses the Kubernetes API to track nodes and pods in your cluster. The metrics-server queries each node over HTTP to fetch metrics. The metrics-server also builds an internal view of pod metadata, and keeps a cache of pod health. That cached pod health information is available via the extension API that the metrics-server makes available.
```

No need to mention that this uses the **watch** verb. If you know what a watch is, you'd guess that metrics-server uses that optimization; if you don't, you might end up wondering what one is.
